Restore custom HPA controller setup

DirectXMan12 opened the pull request:
In previous versions, we had custom setup for the HPA controller,
which allowed the HPA controller to scale DCs, and ensured that it
looked for Heapster in the OpenShift infra namespace.  This was
accidentally removed during a rebase.  This commit restores that logic
using the init functions functionality in the Kube controller manager.


This is currently based on top of #14633

DirectXMan12 commented:
cc @deads2k @liggitt this is the initial shot at what we talked about this morning on IRC

cc @derekwaynecarr this should fix the HPA-related blocker

mfojtik commented:
@DirectXMan12 sorry for yoru suffering :-) the PR you based this on will probably change a lot, can you please open a PR against my branch and I will add it there (and will keep rebasing it to save your pain)

deads2k commented on pkg/cmd/server/kubernetes/master/controller.go:
```diff
@@ -145,3 +149,40 @@ func (c *SchedulerControllerConfig) RunController(ctx kubecontroller.ControllerC
+// to get a handle on OpenShift clients, so that our delegating scales getter can work.
+
+type HorizontalPodAutoscalerControllerConfig struct {
+	HeapsterNamespace string
```
is this really configurable?  why?

DirectXMan12 replied:
I can make it hard-coded.  No particular reason it has to be configurable...

deads2k commented on pkg/cmd/server/kubernetes/master/controller.go:
```diff
@@ -145,3 +149,40 @@ func (c *SchedulerControllerConfig) RunController(ctx kubecontroller.ControllerC
+
+type HorizontalPodAutoscalerControllerConfig struct {
+	HeapsterNamespace string
+	HeapsterScheme    string
```
Why aren't we always https?

deads2k commented on pkg/cmd/server/kubernetes/master/controller.go:
```diff
@@ -145,3 +149,40 @@ func (c *SchedulerControllerConfig) RunController(ctx kubecontroller.ControllerC
+
+func (c *HorizontalPodAutoscalerControllerConfig) RunController(ctx oscontroller.ControllerContext) (bool, error) {
+	hpaClient := ctx.ClientBuilder.ClientOrDie("horizontal-pod-autoscaler")
+	hpaOriginClient := ctx.ClientBuilder.DeprecatedOpenshiftClientOrDie("horizontal-pod-autoscaler")
```
open an issue to switch to the generated client.  I want to delete the old client in 3.7

deads2k commented on pkg/cmd/server/origin/controller.go:
```diff
@@ -143,6 +148,14 @@ func (c *MasterConfig) NewOpenShiftControllerPreStartInitializers() (map[string]
 	// this matches the upstream name
 	ret["serviceaccount-token"] = saToken.RunController
 
+	hpaController := kubernetes.HorizontalPodAutoscalerControllerConfig{
```
Find a way to do this inside of michal's method that's building the entire set of kube init functions.

DirectXMan12 replied:
whoops, I didn't mean to stick this in pre-start -- that was I typo.  I've got a local copy that's moved to the OpenShift controllers initializer area shortly (or was there a different spot that you were thinking?).  It has to be the OpenShift area because we need an OpenShift client handle.

DirectXMan12 commented:
ok, this should work, assuming the base PR gets working (based on my testing, the base PR looks like it's not quite working...)

derekwaynecarr commented:
@DirectXMan12 -- whats missing in this PR is anything that makes me know it won't regress in the future.  what testing can we do to ensure it does not regress in the future by accident?

DirectXMan12 commented:
@derekwaynecarr we need to stand up Heapster as part of our e2e tests, and run the autoscaling e2es.  They might add some time to the e2es, though, since they require waiting for collection of metrics multiple times.

derekwaynecarr commented:
@DirectXMan12 agree we ultimately need that, but i guess what i am wondering is if there is anything short  of that we can do.

DirectXMan12 commented:
not that I can immediately think off... this isn't really something that you can detect outside of e2e, since it's a matter of "is the actually run configuration of the HPA correct when we launch the master"

liggitt commented:
#14633 has merged, can you rebase on master?

DirectXMan12 commented:
rebased, and I've added an initial attempt at a test to ensure that this isn't accidentally broken again.  I'm not entirely happy with it, but basically, it runs the HPA controller from the controller initializes as composed by the master init code.  It passes fake clients to the controller context, and makes sure that the HPA controller calls the right place for DeploymentConfig and Heapster using the fake clients.

DirectXMan12 commented:
cc @mfojtik @deads2k @liggitt does this look good?  What do you think of the test?

DirectXMan12 commented:
(P.S. since this is a blocker, we might just want to merge the fix w/o a regression test, and worry about the regression test later)

liggitt commented on pkg/cmd/server/origin/controller.go:
```diff
@@ -86,6 +86,11 @@ func (c *MasterConfig) NewKubernetesControllerInitalizers(kc *kubernetes.MasterC
+	// NB: this is funky -- it's actually a Kubernetes controller, but we run it as an OpenShift controller in order
+	// to get a handle on OpenShift clients, so that our delegating scales getter can work.  Ergo,
+	// we need to just not do anything right now.
+	delete(ret, "horizontalpodautoscaling")
```
was this necessary? looking at how these are called, openshift ones with the same name take priority. 

I could see adding a check here to make sure "horizontalpodautoscaling" is registered so our override is actually overriding something

liggitt replied:
for example, we override the `serviceaccount` controller without deleting from this list

DirectXMan12 replied:
yeah, looking over again it looks like it's not strictly necessary.  Adding a check to see if it's registered might not be a bad idea

liggitt commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -633,6 +633,27 @@ func (i genericInformers) ForResource(resource schema.GroupVersionResource) (kin
 	return nil, firstErr
 }
 
+func collectInitializers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) (map[string]origincontrollers.InitFunc, error) {
```
revert this change?

DirectXMan12 replied:
yeah, I put it in as part of the stab at a test.  I'm certainly willing to pull all of that out.

sjenning commented:
@DirectXMan12 can you rebase this please? Also, has all the feedback been addressed? Thanks.

DirectXMan12 commented:
@sjenning rebased, pulled out test so that we can evaluate that separately...

liggitt commented on pkg/cmd/server/origin/controller/config.go:
```diff
@@ -60,6 +62,8 @@ func (c *OpenshiftControllerConfig) GetControllerInitializers() (map[string]Init
 	ret["openshift.io/cluster-quota-reconciliation"] = c.ClusterQuotaReconciliationControllerConfig.RunController
 	ret["openshift.io/cluster-quota-mapping"] = c.ClusterQuotaMappingControllerConfig.RunController
 
+	ret["horizontalpodautoscaling"] = c.HorizontalPodAutoscalerControllerConfig.RunController
```
comment that this is overriding the upstream controller

liggitt commented:
nit on comment, with manual testing of scaling both kube and origin resources, this LGTM for 3.6

DirectXMan12 commented:
your nit should be fixed.  Something changed with the RBAC, though.  Tracking it down.

DirectXMan12 commented:
ah... hmm... RBAC is split across two different service accounts: `openshift-infra:horizontal-pod-autoscaler` and `system:controller:horizontal-pod-autoscaler`.  Is that ok (my gut says no, but the client helpers would need to be fiddled to get around it...)?  If so, I just need to restore the RBAC that allows the HPA to access DCs.

@liggitt

DirectXMan12 commented:
alright, this should have the appropriate RBAC, but there's two separate users -- the one from kube (which uses `kube-system:...`) and the one from origin (which uses `openshift-infra:...`).

deads2k commented:
> alright, this should have the appropriate RBAC, but there's two separate users -- the one from kube (which uses kube-system:...) and the one from origin (which uses openshift-infra:...).

Ok, when this happens and we're running an upstream controller, I'd like to keep the identity of the upstream controller but create another role `system:openshift:controller:....` and bind that role to the kube identity.  That way we only have to manage the delta and there are no upstream patches.

DirectXMan12 commented:
Problem: the OpenShift client builder still builds clients using service accounts in the `openshift-infra` namespace, while the `kube` one builds them in the `kube-system` namespace.

Since the OpenShift client builder can build clients for Kube, but not vice-versa, we'd need to use the OpenShift identity and bind the Kube role to it, instead of using the Kube identity and binding the OpenShift role to it.

DirectXMan12 commented:
That sound ok?  We still wouldn't have to manage much, since we're just binding to the kube role, not reinventing it...

deads2k commented on pkg/cmd/server/origin/controller/hpa.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+func (c *HorizontalPodAutoscalerControllerConfig) RunController(ctx ControllerContext) (bool, error) {
+	hpaClient := ctx.KubeControllerContext.ClientBuilder.ClientOrDie(bootstrappolicy.InfraHorizontalPodAutoscalerControllerServiceAccountName)
```
You can get a config from this ClientBuidler

deads2k commented on pkg/cmd/server/origin/controller/hpa.go:
```diff
@@ -0,0 +1,43 @@
+
+func (c *HorizontalPodAutoscalerControllerConfig) RunController(ctx ControllerContext) (bool, error) {
+	hpaClient := ctx.KubeControllerContext.ClientBuilder.ClientOrDie(bootstrappolicy.InfraHorizontalPodAutoscalerControllerServiceAccountName)
+	hpaOriginClient := ctx.ClientBuilder.DeprecatedOpenshiftClientOrDie(bootstrappolicy.InfraHorizontalPodAutoscalerControllerServiceAccountName)
```
And use it to make this client

DirectXMan12 replied:
/me facepalms

Wow, that's the second brainfart I've had while working on this PR... :-/

liggitt commented on pkg/cmd/server/origin/controller/hpa.go:
```diff
@@ -0,0 +1,49 @@
+	hpaClient := ctx.KubeControllerContext.ClientBuilder.ClientOrDie(bootstrappolicy.InfraHorizontalPodAutoscalerControllerServiceAccountName)
+
+	// use the Kubernetes config so that the service account is in the same name namespace for both clients
+	hpaOriginConfig := ctx.KubeControllerContext.ClientBuilder.ConfigOrDie(bootstrappolicy.InfraHorizontalPodAutoscalerControllerServiceAccountName)
```
get the config once, then use it to build the kube and origin clients

deads2k replied:
> get the config once, then use it to build the kube and origin clients

You came out of vacation for *that*? :)

DirectXMan12 replied:
heh.  I almost did that in the last round of revisions, but figured it wasn't worth the extra `if err return err` lines for what should be identical behavior.  I'll add it in.

DirectXMan12 replied:
fixed

derekwaynecarr commented:
[merge][severity:blocker]

DirectXMan12 commented:
ahhhrg.  apparently one of my commits didn't push (it's not serious, just the fix @liggitt mentioned above).

DirectXMan12 commented:
there, *now* it's pushed :-/

derekwaynecarr commented:
[merge][severity:blocker

derekwaynecarr commented:
[merge][severity:blocker]

deads2k commented on pkg/cmd/server/origin/controller/hpa.go:
```diff
@@ -0,0 +1,60 @@
+// NB: this is funky -- it's actually a Kubernetes controller, but we run it as an OpenShift controller in order
+// to get a handle on OpenShift clients, so that our delegating scales getter can work.
+
+type HorizontalPodAutoscalerControllerConfig struct {
```
This belongs under the kube package. It's overriding a kube controller and it has no dependencies on openshift controllercontext information.

deads2k commented on pkg/cmd/server/bootstrappolicy/infra_sa_policy.go:
```diff
@@ -39,6 +39,7 @@ const (
 	InfraServiceIngressIPControllerServiceAccountName           = "service-ingress-ip-controller"
 	InfraPersistentVolumeRecyclerControllerServiceAccountName   = "pv-recycler-controller"
 	InfraResourceQuotaControllerServiceAccountName              = "resourcequota-controller"
+	InfraHorizontalPodAutoscalerControllerServiceAccountName    = "horizontal-pod-autoscaler"
```
move to a separate commented section and indicate that is a special upstream controller that we're adding additional permissions to.

deads2k commented on pkg/cmd/server/origin/controller/config.go:
```diff
@@ -60,6 +62,10 @@ func (c *OpenshiftControllerConfig) GetControllerInitializers() (map[string]Init
 
+	// overrides the Kube HPA controller config, so that we can point it at an HTTPS Heapster
+	// in openshift-infra, and pass it a scale client that knows how to scale DCs
+	ret["horizontalpodautoscaling"] = c.HorizontalPodAutoscalerControllerConfig.RunController
```
This should be done in the kube pacakge of this method.

deads2k commented:
pulling merge tag for comments

DirectXMan12 commented:
ok, moved back to Kube package, and I added a comment on the policy constant.

deads2k commented on pkg/cmd/server/bootstrappolicy/infra_sa_policy.go:
```diff
@@ -38,7 +38,7 @@ const (
 	InfraServiceIngressIPControllerServiceAccountName           = "service-ingress-ip-controller"
 	InfraPersistentVolumeRecyclerControllerServiceAccountName   = "pv-recycler-controller"
-	InfraResourceQuotaControllerServiceAccountName              = "resourcequota-controller"
+	InfraHorizontalPodAutoscalerControllerServiceAccountName    = "horizontal-pod-autoscaler"
```
forgot to move this down to your ocment.

DirectXMan12 replied:
moved the wrong constant ;-)

deads2k commented:
nit on constant location in file. lgtm.

deads2k commented:
[merge]

derekwaynecarr commented:
[merge][severity:blocker]

deads2k commented:
Failures are real

DirectXMan12 commented:
> Failures are real

That's very pseudo zen, @deads2k 

Regenerated the policy fixtures.  I always forget about those.

DirectXMan12 commented:
apparently, the web console test doesn't consider roles starting with `system:openshift:controller:` to be controller roles.  I had to fix that...

eparis commented:
[merge]
